monitor/puller.go

```diff
@@ -31,86 +34,87 @@ func (m *Monitor) processPullerQueue() {
 	iterator.Close()
 	m.lock.RUnlock()
 
+	validators := make(map[mainchain.Addr]bool)
+	delegators := make(map[string]bool)
 	for i, key := range keys {
 		event := NewEventFromBytes(vals[i])
-		log.Infoln("Process puller event", event.Name, "at mainchain block", event.Log.BlockNumber)
+		logmsg := fmt.Sprintf("Process puller event %s at mainchain block %d", event.Name, event.Log.BlockNumber)
 		err = m.dbDelete(key)
 		if err != nil {
-			log.Errorln("db Delete err", err)
+			log.Errorf("%s. db Delete err: %s", logmsg, err)
 			continue
 		}
 
 		switch e := event.ParseEvent(m.EthClient).(type) {
 		case *mainchain.DPoSValidatorChange:
-			m.syncDPoSValidatorChange(e)
+			log.Infof("%s. validator change %x type %d", logmsg, e.EthAddr, e.ChangeType)
+			validators[e.EthAddr] = true
+
+		case *mainchain.DPoSDelegate:
+			log.Infof("%s. delegator %x to candidate %x, stake %s, pool %s", logmsg, e.Delegator, e.Candidate, e.NewStake, e.StakingPool)
+			delegators[delegatorKey(e.Candidate, e.Delegator)] = true
+
 		case *mainchain.DPoSIntendWithdraw:
-			m.syncDPoSIntendWithdraw(e)
+			log.Infof("%s. intend withdraw candidate %x delegator %x amount %s", logmsg, e.Candidate, e.Delegator, e.WithdrawAmount)
+			validators[e.Candidate] = true
+			delegators[delegatorKey(e.Candidate, e.Delegator)] = true
+
 		case *mainchain.DPoSCandidateUnbonded:
-			m.syncDPoSCandidateUnbonded(e)
-		case *mainchain.DPoSConfirmParamProposal:
-			m.syncConfirmParamProposal(e)
+			log.Infof("%s. candidate unbonded %x", logmsg, e.Candidate)
+			validators[e.Candidate] = true
+
 		case *mainchain.DPoSUpdateCommissionRate:
-			m.syncUpdateCommissionRate(e)
+			log.Infof("%s. commission update %x, %s", logmsg, e.Candidate, e.NewRate)
+			validators[e.Candidate] = true
+
 		case *mainchain.SGNUpdateSidechainAddr:
-			m.SyncUpdateSidechainAddr(e.Candidate)
+			m.syncUpdateSidechainAddr(e, logmsg)
+
+		case *mainchain.DPoSConfirmParamProposal:
+			m.syncConfirmParamProposal(e, logmsg)
+
 		case *mainchain.SGNAddSubscriptionBalance:
-			m.syncSGNAddSubscriptionBalance(e)
+			m.syncSGNAddSubscriptionBalance(e, logmsg)
+
 		case *mainchain.CelerLedgerIntendSettle:
 			e.Raw = event.Log
-			m.syncIntendSettle(e)
+			m.syncIntendSettle(e, logmsg)
+
 		case *mainchain.CelerLedgerIntendWithdraw:
 			e.Raw = event.Log
-			m.syncIntendWithdrawChannel(e)
+			m.syncIntendWithdrawChannel(e, logmsg)
 		}
 	}
-}
 
-func (m *Monitor) syncDPoSValidatorChange(validatorChange *mainchain.DPoSValidatorChange) {
-	log.Infof("puller queue process validator change %x", validatorChange.EthAddr)
-	m.SyncValidator(validatorChange.EthAddr)
-}
-
-func (m *Monitor) syncDPoSIntendWithdraw(intendWithdraw *mainchain.DPoSIntendWithdraw) {
-	log.Infof("puller queue process intend withdraw %x", intendWithdraw.Candidate)
-	m.SyncValidator(intendWithdraw.Candidate)
-	m.SyncDelegator(intendWithdraw.Candidate, intendWithdraw.Delegator)
+	for addr := range validators {
```

Comment:
can we have better name for addr and key below

Reply:
done